MVP 기능 구현을 위한 Prisma 스키마 작성 및 마이그레이션 생성

yxxjxxeee opened the pull request:
## #️⃣ 연관된 이슈
- #3

## 📝 작업 내용
> Prisma를 이용해 초기 DB 스키마 작성 및 마이그레이션 생성

## 💬 리뷰 요구사항 (선택)
> 테이블 간 관계 및 구조가 적절하게 설계되었는지 확인 부탁드립니다.

yxxjxxeee committed:
feat: (#3) 초기 Prisma 스키마 및 마이그레이션 생성

dlehdhun commented on prisma/schema.prisma:
```diff
@@ -5,4 +5,180 @@ generator client {
+
+model User {
+  id        Int       @id @default(autoincrement())
+  username  String?   @unique
```
```suggestion
  username String? @db.VarChar(20) @unique
```
Prisma에서 `String`만 사용하면 기본적으로 TEXT 타입으로 매핑됩니다.
그래서 `username`과 같이 길이가 정해져 있는 필드는 `@db.VarChar(20)`나 `@db.VarChar(50)` 식으로 명시해주는 것이 좋을 것 같습니다.

kimjiwonpg98 replied:
nickname을 먼저 등록하는 경우가 있을까요? 없다면 not null로 처리하면 좋을 것 같습니다.

yxxjxxeee replied:
제가 잘 이해한건지 모르겠는데 nickname은 지금도 not null로 처리하고 있고 username은 OAuth로 회원가입한 경우에는 값이 없을 수 있으므로 nullable(String?)로 처리했습니다.

kimjiwonpg98 replied:
헉 username과 password를 not null로 한다는걸 뺴먹고 말했네요
username은 말그대로 이름인걸까요? 유니크 값으로 하면 unique로 했을 때 문제는 없을까요?

yxxjxxeee replied:
username은 로그인할 때 사용하는 아이디입니다!
로그인 방식이 일반 로그인과 OAuth 로그인으로 두 가지인데 OAuth 로그인 계정의 경우 username과 password가 null이기 때문에 이렇게(username과 password를 nullable로) 설계했습니다. 보통 어떻게 하나요?

dlehdhun commented on prisma/schema.prisma:
```diff
@@ -5,4 +5,180 @@ generator client {
+  password  String?
+  nickname  String    @unique
+  createdAt DateTime  @default(now()) @map("created_at")
+  updatedAt DateTime? @map("updated_at")
```
```suggestion
  updatedAt DateTime @updatedAt @map("updated_at")
```
`@updatedAt`을 사용하면 Prisma Client로 update를 실행할 때  자동으로 갱신되므로 편리합니다!
또한 `@updatedAt`는 처음 생성 시 `now()`가 들어가고 이후에는 자동으로 갱신되기 때문에 `?`가 필요 없습니다.

yxxjxxeee replied:
오 생각하지 못했던 부분이네요!

dlehdhun commented on prisma/schema.prisma:
```diff
@@ -5,4 +5,180 @@ generator client {
+  id         Int    @id @default(autoincrement())
+  matchId    Int    @map("match_id")
+  roomUserId Int    @map("room_user_id")
+  imagePath  String @map("image_path")
```
```suggestion
  imagePath  String @db.VarChar(255) @map("image_path")
```
S3 Key만 저장 할 것 같아서 `@db.VarChar(255)`로도 충분 할 것 같습니다!

kimjiwonpg98 commented on prisma/schema.prisma:
```diff
@@ -5,4 +5,180 @@ generator client {
+model User {
+  id        Int       @id @default(autoincrement())
+  username  String?   @unique
+  password  String?
```
password 아마도 hash해서 저장할 것 같은데 아래 Room처럼 네이밍을 명확히 해주면 더 좋을 것 같네요

yxxjxxeee replied:
일관성이 있어서 좋은 것 같습니다!

kimjiwonpg98 commented on prisma/schema.prisma:
```diff
@@ -5,4 +5,180 @@ generator client {
+  @@map("oauth_accounts")
+}
+
+model Room {
```
user 테이블말고는 updatedAt이 없는거같은데 데이터 변경 가능성이 있는 테이블에는 추가되면 좋을 것 같아요

rrgks6221 commented on prisma/schema.prisma:
```diff
@@ -5,4 +5,180 @@ generator client {
+  createdAt       DateTime  @default(now()) @map("created_at")
+  updatedAt       DateTime? @map("updated_at")
+  tutorialDone    Boolean   @default(false) @map("tutorial_done")
+  lastLogin       DateTime? @map("last_login")
```
lastLoginedAt 처럼 시간데이터라는것을 명시하는건 어떻게 생각하시나요?

yxxjxxeee replied:
오 좋은 것 같습니다! 말씀해주신 것처럼 시간 데이터라는 걸 명시적으로 드러내면 더 직관적일 것 같습니다.

rrgks6221 commented on prisma/schema.prisma:
```diff
@@ -5,4 +5,180 @@ generator client {
+  coin            Int       @default(0)
+  createdAt       DateTime  @default(now()) @map("created_at")
+  updatedAt       DateTime? @map("updated_at")
+  tutorialDone    Boolean   @default(false) @map("tutorial_done")
```
튜토리얼 완료 여부의 값을 저장하는것도 좋지만 튜토리얼의 상태를 enum으로 표현하는것도 고려해볼 수 있을거같아요.

ex) null(시작하지 않음), inProgress(진행중), done(완료)

rrgks6221 commented on prisma/schema.prisma:
```diff
@@ -5,4 +5,180 @@ generator client {
+  @@map("rooms")
+}
+
+model RoomUser {
```
보통 특정 공간에 소속된 사용자 Member 도메인으로 많이들 칭합니다.
RDBMS관점에서 봤을 때 관계를 표현하기 위해 테이블은 room_users로 할 수 있겠구요

yxxjxxeee replied:
보통 Member라고 많이 하는군요! 
그럼 Prisma 모델명은 Member로 하고 실제 DB 테이블명은 room_users로 매핑하는 게 좋을까요?

rrgks6221 replied:
추천정도이고 꼭 그렇게 해야한다는 아니예요
이름정도는 팀 내에서 합의가 됐고 다른 사람이 봤을 때 쉽게 이해되게끔만 지어지면 될거같아요

테이블 이름을 말씀드린것은 대부분의 데이터베이스 mapping은 snake_case에 복수형으로 RDBMS 컨벤션에 맞게끔 설정하셔서 Prisma 모델명은 Member로 했더라도 데이터베이스 테이블명은 room_users 로 하는걸 말씀드렸어요

rrgks6221 commented on prisma/schema.prisma:
```diff
@@ -5,4 +5,180 @@ generator client {
+enum Provider {
+  GOOGLE
+  KAKAO
+  NAVER
```
네이버가 웹상에서 우리나라 사람들한테 접근하기 좋은 소셜 로그인 프로바이더이긴 한데 google, kakao와 달리 소셜 로그인을 사용하기 위해서는 심사를 받아야 하는 부분이 있어서 나중에 참고하시면 좋을거같아요

yxxjxxeee replied:
네이버는 심사가 필요하군요 좋은 정보 감사합니다. 참고해서 진행하겠습니다!

rrgks6221 commented on prisma/schema.prisma:
```diff
@@ -5,4 +5,180 @@ generator client {
+  mode         RoomMode
+  maxPlayers   Int       @map("max_players")
+  isPrivate    Boolean   @default(false) @map("is_private")
+  passwordHash String?   @map("password_hash")
```
passwordHash는 패스워드를 해시하기 위한 값처럼 보이는데 hashedPassword는 어떤가요?

yxxjxxeee replied:
좋은 변수명인 것 같습니다!

yxxjxxeee committed:
refactor: (#3) User와 OAuthAccount 모델 통합 및 필드 개선

yxxjxxeee committed:
refactor: (#3) Room 모델 soft delete 및 Match 주제 작성자 필드 추가

- Room에 deletedAt 추가하여 soft delete 지원
- RoomUser를 RoomMember로 이름 변경하여 일관성 확보
- Match에 topic, topicWriterId 필드 추가
- Match.topicWriter를 RoomMember와 관계 설정
- Boolean 필드 네이밍 일관성 정리 (isReady, isTutorialCompleted)

yxxjxxeee committed:
refactor: (#3) 컬럼명 스네이크케이스로 @map 적용
